src/mutable_arithmetics.jl

```diff
@@ -303,12 +308,22 @@ function _MA.add_mul(
         typeof.(args)...,
     )
     expr = _MA.operate(convert, T, lhs)
+    # We can't use `operate!!` here because in the IsNotMutable case (e.g.,
+    # NonlinearExpr), it will fallback to this method and cause a StackOverflow.
+    if _MA.mutability(T) == _MA.IsNotMutable()
```

Comment:
I'm still confused by this but I guess we can fix it later

Reply:
This implementation is the method for `MA.add_mul(...)`, not `MA.operate(::typeof(MA.add_mul), ...)`

Reply:
Look at this:
```julia
julia> import MutableArithmetics as MA

julia> struct Foo <: MA.AbstractMutable
           x::Int
       end

julia> Base.copy(f::Foo) = Foo(f.x)

julia> function MA.add_mul(lhs::Foo, x::Union{Int,Foo}, y::Union{Int,Foo})
           return MA.operate!!(MA.add_mul, copy(lhs), x, y)
       end

julia> lhs = Foo(2)
Foo(2)

julia> x = Foo(3)
Foo(3)

julia> MA.add_mul(lhs, x, 3)
ERROR: StackOverflowError:
Stacktrace:
     [1] operate_fallback!!(::MutableArithmetics.IsNotMutable, ::typeof(MutableArithmetics.add_mul), ::Foo, ::Foo, ::Int64)
       @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/interface.jl:581
     [2] operate!!(::typeof(MutableArithmetics.add_mul), ::Foo, ::Foo, ::Int64)
       @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/rewrite.jl:89
     [3] add_mul(lhs::Foo, x::Foo, y::Int64)
       @ Main ./REPL[26]:2
     [4] operate(op::typeof(MutableArithmetics.add_mul), x::Foo, y::Foo, args::Int64)
       @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/interface.jl:206
     [5] operate_fallback!!(::MutableArithmetics.IsNotMutable, ::typeof(MutableArithmetics.add_mul), ::Foo, ::Foo, ::Int64)
       @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/interface.jl:586
--- the last 4 lines are repeated 19994 more times ---
 [79982] operate!!(::typeof(MutableArithmetics.add_mul), ::Foo, ::Foo, ::Int64)
       @ MutableArithmetics ~/.julia/dev/MutableArithmetics/src/rewrite.jl:89
 [79983] add_mul(lhs::Foo, x::Foo, y::Int64)
       @ Main ./REPL[26]:2
```

`operate!!` falls back to `operate` which falls back to `add_mul` which is this method.